internal/provider/metric_resource.go

```diff
@@ -0,0 +1,357 @@
+	if req.ProviderData == nil {
+		return
+	}
+
+	client, ok := req.ProviderData.(*statuspal.Client)
+
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Data Source Configure Type",
+			fmt.Sprintf(
+				"Expected *statuspal.Client, got: %T. Please report this issue to the provider developers.",
+				req.ProviderData,
+			),
+		)
+
+		return
+	}
+
+	r.client = client
+}
+
+// https://www.statuspal.io/api-docs#tag/Metrics/operation/addMetric
+func (r *MetricResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var data MetricResourceModel
+
+	// Read Terraform plan data into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	if data.Title.IsNull() || data.Title.IsUnknown() {
+		resp.Diagnostics.AddError(
+			"Missing Metric ID",
+			"Cannot create the metric because the ID field is missing from the plan.",
+		)
+
+		return
+	}
+
+	if data.Title.IsNull() || data.Title.IsUnknown() {
+		resp.Diagnostics.AddError(
+			"Missing Metric Title",
+			"Cannot create the metric because the Title field is missing or unknown in the plan.",
+		)
+
+		return
+	}
+
+	if data.Unit.IsNull() || data.Unit.IsUnknown() {
+		resp.Diagnostics.AddError(
+			"Missing Metric Unit",
+			"Cannot create the metric because the Unit field is missing or unknown in the plan.",
+		)
+
+		return
+	}
+
+	model := statuspal.MetricWrite{
+		Title:          data.Title.ValueString(),
+		Unit:           data.Unit.ValueString(),
+		Type:           statuspal.MetricType(data.Type.ValueString()),
+		Enabled:        data.Enabled.ValueBool(),
+		Visible:        data.Visible.ValueBool(),
+		RemoteID:       data.RemoteID.ValueString(),
+		RemoteName:     data.RemoteName.ValueString(),
+		Threshold:      data.Threshold.ValueInt64(),
+		FeaturedNumber: statuspal.FeaturedNumber(data.FeaturedNumber.ValueString()),
+		IntegrationID:  data.IntegrationID.ValueString(),
+	}
+
+	metric, err := r.client.CreateMetric(data.Subdomain.String(), &model)
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to create the metric, got error: %s", err))
+
+		return
+	}
+
+	data.ID = types.Int64Value(metric.ID)
+	data.Enabled = types.BoolValue(metric.Enabled)
+	data.Title = types.StringValue(metric.Title)
+	data.Unit = types.StringValue(metric.Unit)
+	data.Type = types.StringValue(string(metric.Type))
+	data.Visible = types.BoolValue(metric.Visible)
+	data.RemoteID = types.StringValue(metric.RemoteID)
+	data.RemoteName = types.StringValue(metric.RemoteName)
+	data.Status = types.StringValue(metric.Status)
+	data.LatestEntryTime = types.Int64Value(metric.LatestEntryTime)
+	data.Threshold = types.Int64Value(metric.Threshold)
+	data.FeaturedNumber = types.StringValue(string(metric.FeaturedNumber))
+	data.Order = types.Int64Value(metric.Order)
+	data.IntegrationID = types.StringValue(metric.IntegrationID)
```

Comment:
Can we use "mapper" functions here to minimize the code duplication.
See the service resource:
https://github.com/statuspal/terraform-provider-statuspal/blob/e1c48870df86da38c85828e706c1cf784ce1cb62/internal/provider/service_resource.go#L399
https://github.com/statuspal/terraform-provider-statuspal/blob/e1c48870df86da38c85828e706c1cf784ce1cb62/internal/provider/service_resource.go#L416